kythe/cxx/extractor/proto/proto_extractor.cc

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "proto_extractor.h"
+
+#include <string>
+
+#include "absl/strings/str_cat.h"
+#include "glog/logging.h"
+#include "google/protobuf/compiler/importer.h"
+#include "google/protobuf/io/zero_copy_stream_impl.h"
+#include "kythe/cxx/common/file_vname_generator.h"
+#include "kythe/cxx/common/path_utils.h"
+#include "kythe/cxx/indexer/proto/search_path.h"
+#include "kythe/proto/analysis.pb.h"
+
+namespace kythe {
+namespace lang_proto {
+namespace {
+
+// Error "collector" that just writes messages to log output.
+class LoggingMultiFileErrorCollector
+    : public google::protobuf::compiler::MultiFileErrorCollector {
+ public:
+  void AddError(const std::string& filename, int line, int column,
+                const std::string& message) override {
+    LOG(ERROR) << filename << "@" << line << ":" << column << ": " << message;
+  }
+
+  void AddWarning(const std::string& filename, int line, int column,
+                  const std::string& message) override {
+    LOG(ERROR) << filename << "@" << line << ":" << column << ": " << message;
+  }
+};
+
+// DiskSourceTree that records which proto files are opened while parsing the
+// toplevel proto(s), allowing us to get a list of transitive dependencies.
+class RecordingDiskSourceTree
+    : public google::protobuf::compiler::DiskSourceTree {
+ public:
+  google::protobuf::io::ZeroCopyInputStream* Open(
+      const std::string& filename) override {
+    // Record resolved/canonical path because the same proto may be Open()'d via
+    // multiple relative paths and we only want to record it once.
+    std::string canonical_path;
+    if (!DiskSourceTree::VirtualFileToDiskFile(filename, &canonical_path)) {
+      return nullptr;
+    }
+    if (opened_files_.find(canonical_path) == opened_files_.end()) {
+      opened_files_.insert(canonical_path);
+    }
+
+    return DiskSourceTree::Open(filename);
+  }
+
+  // A set of unique file paths that have been passed to Open().
+  const std::set<std::string>& opened_files() const { return opened_files_; }
```

Comment:
It looks a lot like the C++ files need to be run through clang-format (or those of us with the suggested commit-hooks set up will get warning/errors).

Reply:
I just ran clang-format and it only found a couple things in proto_indexer.cc and index_pack_test.cc. We're using the config at `kythe/.clang-format`, right?

Reply:
Yup, that's right.  This line looked a bit long, but I miscounted :-)